pkg/client/client_v2_test.go

```diff
@@ -0,0 +1,696 @@
+					unitsLock.Unlock()
+					change.Unit.UpdateState(UnitStateHealthy, "Healthy", map[string]interface{}{
+						"custom": "payload",
+					})
+				case UnitChangedModified:
+					state, _ := change.Unit.Expected()
+					if state == UnitStateStopped {
+						change.Unit.UpdateState(UnitStateStopping, "Stopping", nil)
+						go func() {
+							time.Sleep(100 * time.Millisecond)
+							change.Unit.UpdateState(UnitStateStopped, "Stopped", nil)
+						}()
+					}
+				case UnitChangedRemoved:
+					unitsLock.Lock()
+					delete(units, change.Unit.ID())
+					unitsLock.Unlock()
+				}
+
+			}
+		}
+	}()
+
+	require.NoError(t, client.Start(ctx))
+	defer client.Stop()
+	require.NoError(t, waitFor(func() error {
+		m.Lock()
+		defer m.Unlock()
+
+		if !connected {
+			return fmt.Errorf("server never recieved valid token")
+		}
+		return nil
+	}))
+	require.NoError(t, waitFor(func() error {
+		m.Lock()
+		defer m.Unlock()
+
+		if unitTwo.state != UnitStateStopped {
+			return fmt.Errorf("never unitTwo into stopped state")
+		}
+		return nil
+	}))
+
+	unitsLock.Lock()
+	defer unitsLock.Unlock()
+	assert.Len(t, units, 1)
+	assert.Equal(t, UnitStateHealthy, unitOne.state)
+	assert.Equal(t, "Healthy", unitOne.stateMsg)
+	assert.Equal(t, UnitStateStopped, unitTwo.state)
+	assert.Equal(t, "Stopped", unitTwo.stateMsg)
+}
+
+func TestClientV2_Actions(t *testing.T) {
+	var m sync.Mutex
+	token := newID()
+	gotInit := false
+	srv := StubServerV2{
+		CheckinV2Impl: func(observed *proto.CheckinObserved) *proto.CheckinExpected {
+			if observed.Token == token {
+				return &proto.CheckinExpected{
+					Units: []*proto.UnitExpected{
+						{
+							Id:             newID(),
+							Type:           proto.UnitType_OUTPUT,
+							State:          proto.State_HEALTHY,
+							ConfigStateIdx: 1,
+							Config:         "config",
+						},
+					},
+				}
+			}
+			// disconnect
+			return nil
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			m.Lock()
+			defer m.Unlock()
+
+			if response.Token != token {
+				return fmt.Errorf("invalid token")
+			}
+			if response.Id == "init" {
+				gotInit = true
+			}
+			return nil
+		},
+		actions:     make(chan *performAction, 100),
+		sentActions: make(map[string]*performAction),
+	}
+	require.NoError(t, srv.Start())
+	defer srv.Stop()
+
+	var errsLock sync.Mutex
+	var errs []error
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := NewV2(fmt.Sprintf(":%d", srv.Port), token, VersionInfo{}, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	storeErrors(ctx, client, &errs, &errsLock)
+
+	var unitsLock sync.Mutex
+	var units []*Unit
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case change := <-client.UnitChanges():
+				switch change.Type {
+				case UnitChangedAdded:
+					unitsLock.Lock()
+					units = append(units, change.Unit)
+					unitsLock.Unlock()
+				default:
+					panic("not implemented")
```

Comment:
let's use testing's `t.Fatal`

Reply:
Its in a seperate goroutine, `go vet` will not allow this.

`call to (*T).Fatal from a non-test goroutine`